blockchain/treasury.go

```diff
@@ -730,6 +730,58 @@ func (b *BlockChain) maxTreasuryExpenditureDCP0006(preTVINode *blockNode) (int64
 	return allowedToSpend, nil
 }
 
+// maxTreasuryExpenditureDCP0007 returns the maximum amount of funds that can
+// be spent from the treasury at the block after the provided node.
+//
+// This is code that was activated as part of the 'reverttreasurypolicy'
+// consensus upgrade as defined in DCP0007.
+//
+// See notes on maxTreasuryExpenditure().
+func (b *BlockChain) maxTreasuryExpenditureDCP0007(preTVINode *blockNode) (int64, error) {
+	// The expenditure policy check is roughly speaking:
+	//
+	//     "The sum of tspends inside an expenditure window cannot exceed
+	//     the total amount of income received by the treasury in the same
+	//     window in addition to an X% increase."
+	//
+	// Currently this function is pretty naive. It simply iterates through
+	// prior blocks one at a time. This is very expensive and may need to
+	// be rethought into a proper index.
+
+	policyWindow := b.chainParams.TreasuryVoteInterval *
+		b.chainParams.TreasuryVoteIntervalMultiplier *
+		b.chainParams.TreasuryExpenditureWindow
+
+	// Each policy window starts at a TVI block and ends at the block
+	// immediately prior to another TVI (inclusive of preTVINode).
+	//
+	// First: sum up tspends, tadds and tbases inside the most recent
+	// policyWindow.
+	spentRecent, addedRecent, _, err := b.sumPastTreasuryChanges(preTVINode, policyWindow)
+	if err != nil {
+		return 0, err
+	}
+
+	// Treasury can spend up to 150% the amount received in the previous
+	// window.
+	addedPlusAllowance := addedRecent + addedRecent/2
+
+	// The maximum expenditure allowed for the next block is the difference
+	// between the maximum possible and what has already been spent in the most
+	// recent policy window. This is capped at zero on the lower end to account
+	// for cases where the policy _already_ spent more than the allowed.
+	var allowedToSpend int64
+	if addedPlusAllowance > spentRecent {
+		allowedToSpend = addedPlusAllowance - spentRecent
+	}
+
+	trsyLog.Tracef("  maxTSpendExpenditureDCP0007: spent %d, "+
```

Comment:
I won't hold this PR up over it, because I know it's following suit with the other logs in the treasury code, but, just noting that I really dislike the way the treasury bits in general do their logging like this.  Imo, this is something that a _caller_ should be choosing whether to log or not.  Otherwise, it ends up logging every time it's called for no reason.  Moreover, the addition of the extra spaces at the beginning make them all inconsistent with the rest of the code.

Anyway, as I said, it's not something I'll hold up this PR over, but, at some point, I'm very likely going to be ripping all these non-standard log statement out.